### PR TITLE
Add aws session id metadata

### PR DESCRIPTION
**Pull request by forty**
Trying to implement #5761 
I tried my best with kind help from @jefferai , but so far I did not manage to get the token metadata to path_sign.go
If anyone has an idea on what's wrong with my code, any suggestion would be appreciated
Thanks

### CODE REVIEW

File: builtin/credential/aws/path_login.go [hunk 1]
```diff
@@ -1314,6 +1320,7 @@ func (b *backend) pathLoginUpdateIam(ctx context.Context, req *logical.Request,
 				"client_arn":           callerID.Arn,
 				"canonical_arn":        entity.canonicalArn(),
 				"client_user_id":       callerUniqueId,
+				"client_session_id":    callerSessionId,
```
Comment (jefferai) [inline builtin/credential/aws/path_login.go]: This is token metadata, but....

File: builtin/logical/ssh/path_sign.go [hunk 2]
```diff
@@ -285,11 +285,21 @@ func (b *backend) calculateKeyId(data *framework.FieldData, req *logical.Request
...
 		"role_name":          data.Get("role").(string),
 		"public_key_hash":    fmt.Sprintf("%x", sha256.Sum256(pubKey.Marshal())),
-	})
+	}
+
+	entity, _ := b.System().EntityInfo(req.EntityID)
+
+	if entity != nil {
+		for key, value := range entity.Metadata {
+			substMap[fmt.Sprintf("token_meta_%s", key)] = value
```
Comment (jefferai) [inline builtin/logical/ssh/path_sign.go]: ...this is entity metadata, not token metadata.

You'd want to put the metadata into the Alias info that is returned from AWS, then access that Alias here. Except a challenge is that you don't really know what alias it is, so you can't do this just from code; you'd need some kind of parameter that can be set on the role to indicate what alias it should source key info from, and which keys from that alias' metadata.

Comment (forty) [reply builtin/logical/ssh/path_sign.go]: Thanks, I'll look into Alias, looks more challenging than what I expected, but I'll try

Comment (forty) [reply builtin/logical/ssh/path_sign.go]: @jefferai So I have something working now, with very simple alias selection rule: it takes the first one if any. Would it be enough to merge it as is? I'm not sure whether having several entity aliases is a typical use case?
I could add more criteria to filter the aliases, but I don't think I can be sure only one alias will match. If we just wand to add a filter, I think it would make more sense to have a filter on either the alias mount_path, the alias mount_type or maybe on a pattern of the alias name (in any case I would have to take the first one matching the condition)
What do you think?

Comment (forty) [reply builtin/logical/ssh/path_sign.go]: @jefferai congrats on the release of vault 1.0 :) kind bump for my questions above :)

**Comment by forty** [comment]
Hi @jefferai , happy new year!

I'm copying my latest comment here as it's currently on a commit I dropped, which might not be the most convenient place to have a discussion:

So I have something working now, with very simple alias selection rule: it takes the first one if any. Would it be enough to merge it as is? I'm not sure whether having several entity aliases is a typical use case?
I could add more criteria to filter the aliases, but I don't think I can be sure only one alias will match. If we just wand to add a filter, I think it would make more sense to have a filter on either the alias mount_path, the alias mount_type or maybe on a pattern of the alias name (in any case I would have to take the first one matching the condition)
What do you think?

Thanks again for your help on this

**Comment by forty** [comment]
> > it takes the first one if any
> 
> First one of what? Which value in the metadata are you referring to?

@jefferai The first entity alias. The metadata are linked to an alias, but I don't know which alias was used. See my code in path_sign: I do `b.System().EntityInfo(req.EntityID).Aliases[0].GetMetadata()`, so taking the first alias of the list

**Comment by tidalf** [comment]
Hello @jefferai, 

I'd love to see this integrated (without it we don't have a proper way to identify who is using the ssh cert signing feature using aws roles). Is it worth to rebase/fix the tests or there is no way it can be marged ?

Thanks in advance !
Cyril

**Comment by jefferai** [comment]
At the moment we cannot progress on it because the CLA has not been signed.

**Comment by tidalf** [comment]
> At the moment we cannot progress on it because the CLA has not been signed.
Thanks, it is now :)

File: builtin/credential/aws/path_login.go [hunk 3]
```diff
@@ -1193,7 +1193,13 @@ func (b *backend) pathLoginUpdateIam(ctx context.Context, req *logical.Request,
 	}
 	// This could either be a "userID:SessionID" (in the case of an assumed role) or just a "userID"
 	// (in the case of an IAM user).
-	callerUniqueId := strings.Split(callerID.UserId, ":")[0]
+	splittedUserId := strings.Split(callerID.UserId, ":")
```
Comment (jefferai) [inline builtin/credential/aws/path_login.go]:
```suggestion
	splitUserId := strings.Split(callerID.UserId, ":")
```

File: builtin/logical/ssh/path_sign.go [hunk 4]
```diff
@@ -285,11 +285,21 @@ func (b *backend) calculateKeyId(data *framework.FieldData, req *logical.Request
 		keyIDFormat = role.KeyIDFormat
 	}
 
-	keyID := substQuery(keyIDFormat, map[string]string{
+	substMap := map[string]string{
 		"token_display_name": req.DisplayName,
 		"role_name":          data.Get("role").(string),
 		"public_key_hash":    fmt.Sprintf("%x", sha256.Sum256(pubKey.Marshal())),
-	})
+	}
```
Comment (jefferai) [inline builtin/logical/ssh/path_sign.go]: Please don't mix changes into one PR -- this needs to be a separate PR.

**Comment by tyrannosaurus-becks** [comment]
Closing this due to lack of activity. This work is much appreciated, though! Please do feel free to reopen this if there's a chance to circle back and follow up. Thank you!
